File: scripts/build_codebook.jl
#!/usr/bin/env julia
# Generate codebook.md for every dataset from its datapackage.json.
#
# Julia equivalent of build_codebook.py. One source of truth: the schema; the
# codebook is compiled, never hand-edited. Depends only on JSON.jl (Julia ships
# no JSON parser in the standard library); the CSV coverage count is done with a
# small quote-aware parser so no CSV package is required.
#
# Choose ONE generator for the repo — Python OR Julia — not both. Two generators
# that must be kept in sync is exactly the maintenance liability the project is
# trying to avoid. See CONTRIBUTING.md.
#
# Usage:
#   julia scripts/build_codebook.jl          # regenerate all
#   julia scripts/build_codebook.jl <dir>    # regenerate one dataset folder

import JSON

const ROOT = normpath(joinpath(@__DIR__, ".."))
const DATASETS = joinpath(ROOT, "datasets")

"Parse one CSV line into fields, honouring double-quoted fields and \"\" escapes."
function parse_csv_line(line::AbstractString)
    fields = String[]
    buf = IOBuffer()
    inquote = false
    i = firstindex(line)
    while i <= lastindex(line)
        c = line[i]
        if inquote
            if c == '"'
                if i < lastindex(line) && line[nextind(line, i)] == '"'
                    write(buf, '"'); i = nextind(line, i)
                else
                    inquote = false
                end
            else
                write(buf, c)
            end
        else
            if c == '"'
                inquote = true
            elseif c == ','
                push!(fields, String(take!(buf)))
            else
                write(buf, c)
            end
        end
        i = nextind(line, i)
    end
    push!(fields, String(take!(buf)))
    return fields
end

"Return (n_rows, present::Dict) counting non-missing cells per field."
function coverage(csv_path, fields, missing_tokens)
    present = Dict{String,Int}(f["name"] => 0 for f in fields)
    isfile(csv_path) || return (0, present)
    miss = Set(String.(missing_tokens))
    lines = readlines(csv_path)
    isempty(lines) && return (0, present)
    header = parse_csv_line(lines[1])
    idx = Dict(name => k for (k, name) in enumerate(header))
    n = 0
    for line in lines[2:end]
        isempty(strip(line)) && continue
        n += 1
        cells = parse_csv_line(line)
        for f in fields
            k = get(idx, f["name"], 0)
            k == 0 && continue
            v = k <= length(cells) ? strip(cells[k]) : ""
            if v != "" && !(v in miss)
                present[f["name"]] += 1
            end
        end
    end
    return (n, present)
end

md_escape(s) = replace(replace(something(s, ""), "|" => "\\|"), "\n" => " ")

function build(dataset_dir)
    dp = JSON.parsefile(joinpath(dataset_dir, "datapackage.json"))
    res = dp["resources"][1]
    schema = res["schema"]
    fields = schema["fields"]
    missing_tokens = get(schema, "missingValues", [""])
    csv_path = joinpath(dataset_dir, res["path"])
    n_rows, present = coverage(csv_path, fields, missing_tokens)

    lic = join([get(l, "name", "") for l in get(dp, "licenses", [])], ", ")
    lic = isempty(lic) ? "—" : lic
    contribs = join([string(get(c, "title", ""), " (", get(c, "role", ""), ")")
                     for c in get(dp, "contributors", [])], ", ")
    contribs = isempty(contribs) ? "—" : contribs

    io = IOBuffer()
    w(s="") = println(io, s)

    w("# Codebook — $(get(dp, "title", dp["name"]))\n")
    w("> **Generated file.** Do not edit by hand. Produced by " *
      "`scripts/build_codebook.jl` from `datapackage.json`. " *
      "Edit the schema and regenerate.\n")
    w("- **Dataset**: `$(dp["name"])`  ")
    w("- **Version**: $(get(dp, "version", "—"))  ")
    w("- **License**: $lic  ")
    w("- **Contributors**: $contribs  ")
    w("- **Rows**: $n_rows  ")
    w("- **Generated**: deterministically from `datapackage.json` (timestamps via Git history)\n")
    w("\n$(get(dp, "description", ""))\n")
    w("\n## Provenance\n")
    w("Attribution and timestamps are supplied by Git (`git blame` for line-level " *
      "history); releases are frozen and citable via a Zenodo DOI. Per-observation " *
      "coder attribution is carried in the `coder` field.\n")
    w("\n## Missing-value conventions\n")
    w("Absence is coded, never blank. These tokens are treated as missing by the " *
      "schema (`missingValues`):\n")
    w("\n| Token | Meaning |")
    w("|---|---|")
    w("| `.NR` | not recorded in the source |")
    w("| `.IL` | present but illegible / damaged |")
    w("| `.NA` | not applicable to this record type |")
    w("\n> `.ZERO` is **not** here: a source-recorded zero is the value `0`, a datum, " *
      "not an absence (see the `missingness` field).\n")
    w("\n## Variables at a glance\n")
    w("| # | Field | Type | Required | Coded values | Present |")
    w("|---:|---|---|:---:|---|---:|")
    for (i, f) in enumerate(fields)
        c = get(f, "constraints", Dict())
        req = get(c, "required", false) ? "✓" : ""
        enum = haskey(c, "enum") ? join(["`$v`" for v in c["enum"]], ", ") : ""
        cov = n_rows > 0 ? "$(present[f["name"]])/$n_rows" : "—"
        w("| $i | `$(f["name"])` | $(get(f, "type", "")) | $req | $(md_escape(enum)) | $cov |")
    end
    w("\n## Variable definitions\n")
    for f in fields
        c = get(f, "constraints", Dict())
        w("\n### `$(f["name"])` — $(md_escape(get(f, "title", "")))\n")
        w(md_escape(get(f, "description", "")) * "\n")
        bits = ["**type** $(get(f, "type", ""))"]
        get(c, "required", false) && push!(bits, "**required**")
        get(c, "unique", false) && push!(bits, "**unique**")
        haskey(c, "pattern") && push!(bits, "**pattern** `$(c["pattern"])`")
        haskey(c, "enum") && push!(bits, "**values** " * join(["`$v`" for v in c["enum"]], ", "))
        w("- " * join(bits, " · ") * "\n")
    end

    dest = joinpath(dataset_dir, "codebook.md")
    write(dest, String(take!(io)))
    return dest
end

function main()
    targets = if length(ARGS) >= 1
        [ARGS[1]]
    else
        sort([joinpath(DATASETS, d) for d in readdir(DATASETS)
              if isfile(joinpath(DATASETS, d, "datapackage.json"))])
    end
    for d in targets
        dest = build(d)
        println("wrote ", relpath(dest, ROOT))
    end
end

main()

File: scripts/build_codebook.py
#!/usr/bin/env python3
"""Generate codebook.md for every dataset from its datapackage.json.

One source of truth: the schema. The human-readable codebook is compiled, never
hand-edited. Standard library only, so it runs anywhere with no install.

Usage:
    python scripts/build_codebook.py          # regenerate all
    python scripts/build_codebook.py <dir>    # regenerate one dataset folder
"""
from __future__ import annotations

import csv
import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
DATASETS = ROOT / "datasets"


def coverage(csv_path: Path, fields: list[dict], missing: list[str]) -> tuple[int, dict[str, int]]:
    """Count rows and, per field, how many are present (not a missing token)."""
    miss = set(missing)
    present: dict[str, int] = {f["name"]: 0 for f in fields}
    rows = 0
    if not csv_path.exists():
        return 0, present
    with csv_path.open(encoding="utf-8") as fh:
        for row in csv.DictReader(fh):
            rows += 1
            for f in fields:
                v = (row.get(f["name"]) or "").strip()
                if v != "" and v not in miss:
                    present[f["name"]] += 1
    return rows, present


def md_escape(text: str) -> str:
    return (text or "").replace("|", "\\|").replace("\n", " ")


def md_table(headers: list[str], rows: list[list[str]], aligns: str = "") -> list[str]:
    """Render an aligned GitHub-flavoured markdown table.

    aligns: one char per column, 'l' (default) / 'c' / 'r'.
    """
    aligns = (aligns + "l" * len(headers))[: len(headers)]
    widths = [len(h) for h in headers]
    for row in rows:
        for i, c in enumerate(row):
            widths[i] = max(widths[i], len(c))

    def cell(text: str, w: int, a: str) -> str:
        return text.rjust(w) if a == "r" else text.center(w) if a == "c" else text.ljust(w)

    def sep(w: int, a: str) -> str:
        if a == "r":
            return "-" * (w + 1) + ":"
        if a == "c":
            return ":" + "-" * w + ":"
        return "-" * (w + 2)

    lines = ["| " + " | ".join(cell(h, w, a) for h, w, a in zip(headers, widths, aligns)) + " |"]
    lines.append("|" + "|".join(sep(w, a) for w, a in zip(widths, aligns)) + "|")
    for row in rows:
        lines.append("| " + " | ".join(cell(c, w, a) for c, w, a in zip(row, widths, aligns)) + " |")
    return lines


def build(dataset_dir: Path) -> Path:
    dp = json.loads((dataset_dir / "datapackage.json").read_text(encoding="utf-8"))
    res = dp["resources"][0]
    schema = res["schema"]
    fields = schema["fields"]
    missing = schema.get("missingValues", [""])
    csv_path = dataset_dir / res["path"]
    n_rows, present = coverage(csv_path, fields, missing)

    lic = ", ".join(l.get("name", "") for l in dp.get("licenses", [])) or "—"
    contribs = ", ".join(
        f'{c.get("title","")} ({c.get("role","")})' for c in dp.get("contributors", [])
    ) or "—"

    out = []
    w = out.append
    w(f"# Codebook — {dp.get('title', dp['name'])}\n")
    w("> **Generated file.** Do not edit by hand. Produced by "
      "`scripts/build_codebook.py` from `datapackage.json`. "
      "Edit the schema and regenerate.\n")
    w(f"- **Dataset**: `{dp['name']}`  ")
    w(f"- **Version**: {dp.get('version','—')}  ")
    w(f"- **License**: {lic}  ")
    w(f"- **Contributors**: {contribs}  ")
    w(f"- **Rows**: {n_rows}  ")
    w("- **Generated**: deterministically from `datapackage.json` "
      "(timestamps via Git history)\n")
    w(f"\n{dp.get('description','')}\n")

    w("\n## Provenance\n")
    w("Attribution and timestamps are supplied by Git (`git blame` for line-level "
      "history); releases are frozen and citable via a FigShare and Zenodo DOI. "
      "Per-observation coder attribution is carried in the `coder` field.\n")

    w("\n## Missing-value conventions\n")
    w("Absence is coded, never blank. These tokens are treated as missing by the "
      "schema (`missingValues`):\n")
    for line in md_table(
        ["Token", "Meaning"],
        [
            ["`.NR`", "not recorded in the source"],
            ["`.IL`", "present but illegible / damaged"],
            ["`.NA`", "not applicable to this record type"],
        ],
    ):
        w(line)
    w("\n> `.ZERO` is **not** here: a source-recorded zero is the value `0`, a datum, "
      "not an absence (see the `missingness` field).\n")

    w("\n## Variables at a glance\n")
    rows = []
    for i, f in enumerate(fields, 1):
        c = f.get("constraints", {}) or {}
        req = "✓" if c.get("required") else ""
        enum = ", ".join(f"`{v}`" for v in c["enum"]) if "enum" in c else ""
        cov = f"{present.get(f['name'], 0)}/{n_rows}" if n_rows else "—"
        rows.append([str(i), f"`{f['name']}`", f.get("type", ""), req, md_escape(enum), cov])
    for line in md_table(["#", "Field", "Type", "Required", "Coded values", "Present"], rows, aligns="rllclr"):
        w(line)

    w("\n## Variable definitions\n")
    for f in fields:
        c = f.get("constraints", {}) or {}
        w(f"\n### `{f['name']}` — {md_escape(f.get('title',''))}\n")
        w(f"{md_escape(f.get('description',''))}\n")
        bits = [f"**type** {f.get('type','')}"]
        if c.get("required"):
            bits.append("**required**")
        if c.get("unique"):
            bits.append("**unique**")
        if c.get("pattern"):
            bits.append(f"**pattern** `{c['pattern']}`")
        if "enum" in c:
            bits.append("**values** " + ", ".join(f"`{v}`" for v in c["enum"]))
        w("- " + " · ".join(bits) + "\n")

    text = "\n".join(out) + "\n"
    dest = dataset_dir / "codebook.md"
    dest.write_text(text, encoding="utf-8")
    return dest


def main() -> None:
    targets = (
        [Path(sys.argv[1])]
        if len(sys.argv) > 1
        else sorted(
            p for p in DATASETS.iterdir()
            if (p / "datapackage.json").exists() and not p.name.startswith("_")
        )
    )
    for d in targets:
        dest = build(d)
        print(f"wrote {dest.relative_to(ROOT)}")


if __name__ == "__main__":
    main()
